src/sys/windows/named_pipe.rs

```diff
@@ -0,0 +1,610 @@
+            // In theory not possible with `ready_registration` checked above,
+            // but return would block for now.
+            State::None => {
+                state.read_waker = Some(cx.waker().clone());
+                Poll::Pending
+            }
+
+            // A read is in flight, still waiting for it to finish
+            State::Pending(buf, amt) => {
+                state.read = State::Pending(buf, amt);
+                state.read_waker = Some(cx.waker().clone());
+                Poll::Pending
+            }
+
+            // We previously read something into `data`, try to copy out some
+            // data. If we copy out all the data schedule a new read and
+            // otherwise store the buffer to get read later.
+            State::Ok(data, cur) => {
+                let n = {
+                    let mut remaining = &data[cur..];
+                    remaining.read(buf)?
+                };
+                let next = cur + n;
+                if next != data.len() {
+                    state.read = State::Ok(data, next);
+                } else {
+                    self.inner.put_buffer(data);
+                    Inner::schedule_read(&self.inner, &mut state);
+                }
+                Poll::Ready(Ok(n))
+            }
+
+            // Looks like an in-flight read hit an error, return that here while
+            // we schedule a new one.
+            State::Err(e) => {
+                Inner::schedule_read(&self.inner, &mut state);
+                if e.raw_os_error() == Some(ERROR_BROKEN_PIPE as i32) {
+                    Poll::Ready(Ok(0))
+                } else {
+                    Poll::Ready(Err(e))
+                }
+            }
+        }
+    }
+
+    /// TODO: dox
+    pub fn write(&mut self, cx: &mut Context<'_>, buf: &[u8]) -> Poll<io::Result<usize>> {        
+        // Make sure there's no writes pending
+        let mut io = self.inner.io.lock().unwrap();
+        match io.write {
+            State::None => {}
+            _ => {
+                io.write_waker = Some(cx.waker().clone());
+                return Poll::Pending;
+            }
+        }
+
+        // Move `buf` onto the heap and fire off the write
+        let mut owned_buf = self.inner.get_buffer();
+        owned_buf.extend(buf);
+        Inner::schedule_write(&self.inner, owned_buf, 0, &mut io);
+        Poll::Ready(Ok(buf.len()))
+    }
+}
+
+impl AsRawHandle for NamedPipe {
+    fn as_raw_handle(&self) -> RawHandle {
+        self.inner.handle.as_raw_handle()
+    }
+}
+
+impl fmt::Debug for NamedPipe {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.inner.handle.fmt(f)
+    }
+}
+
+impl Drop for NamedPipe {
+    fn drop(&mut self) {
+        // Cancel pending reads/connects, but don't cancel writes to ensure that
+        // everything is flushed out.
+        unsafe {
+            if self.inner.connecting.load(SeqCst) {
```

Comment:
Nit: this can be lowered to `AcqRel`.

Reply:
I don't believe this is true. Switching to `AcqRel` permits it to load stale values, i.e. we can see values that are not the "newest".